### ISSUE TITLE
thunder.jit of AutoEncoder in NeMo Stable Diffusion slower than eager

### ISSUE THREAD

**Issue by athitten** [Lightning-AI/lightning-thunder#468]
### 🐛 Bug

Applying `thunder.jit` to the `AutoEncoder` stage in NeMo's Stable Diffusion is slower than the eager mode (takes  0.198s per train step) where as `thunder.jit` 0.242s takes per step.

### To Reproduce

1. Apply the attached git diff to NeMo
[encoder.patch](https://github.com/Lightning-AI/lightning-thunder/files/15478010/encoder.patch)
[encoder.patch](https://github.com/Lightning-AI/lightning-thunder/files/15478010/encoder.patch)
[encoder.patch](https://github.com/Lightning-AI/lightning-thunder/files/15478010/encoder.patch)

2. Run NeMo using the command below:
```python
python examples/multimodal/text_to_image/stable_diffusion/sd_train.py trainer.precision=16 trainer.num_nodes=1 trainer.devices=1 ++exp_manager.max_time_per_run=00:00:03:00 trainer.max_steps=20 model.micro_batch_size=1 model.global_batch_size=1 model.data.synthetic_data=True exp_manager.exp_dir=/workspace/TestData/multimodal/stable_diffusion_train model.inductor=False model.cond_stage_config._target_=nemo.collections.multimodal.modules.stable_diffusion.encoders.modules.FrozenCLIPEmbedder ++model.cond_stage_config.version=openai/clip-vit-large-patch14 ++model.cond_stage_config.max_length=77 ~model.cond_stage_config.restore_from_path ~model.cond_stage_config.freeze ~model.cond_stage_config.layer model.unet_config.from_pretrained=null model.first_stage_config.from_pretrained=null model.unet_config.use_flash_attention=False model.unet_config.attention_resolutions=\[1\] model.unet_config.channel_mult=\[1\]
```
The last trace of the encode step can be found in the attached log.
[last_trace_SD_encoder.log](https://github.com/Lightning-AI/lightning-thunder/files/15477773/last_trace_SD_encoder.log)
[last_trace_SD_encoder.log](https://github.com/Lightning-AI/lightning-thunder/files/15477773/last_trace_SD_encoder.log)
[last_trace_SD_encoder.log](https://github.com/Lightning-AI/lightning-thunder/files/15477773/last_trace_SD_encoder.log)

**Comment by tfogal** [comment]
@parthmannan we could use some help here. Can you help us identify why we're causing slowdowns?

**Comment by parthmannan** [comment]
Yep, I just need a couple days to wrap an urgent task I have been occupied with and can take a look at this early next week.

**Comment by tfogal** [comment]
convo w/ Abhishree:
* There is a `last_traces` log attached here!
* Should look into the ordering of ops--are the nvFusion* regions being produced properly?
* There are some permutes---do we really need these? Is this an impact of our current (lack of) layout algorithm?

**Comment by parthmannan** [comment]
>There are some permutes---do we really need these? Is this an impact of our current (lack of) layout algorithm?

I am yet to run this myself but the permutes seem okay. The permute seems to be happening before a `BMM` layer and one of the tensors would need to be permuted to perform the op.
However, I am not sure if this could just be a `view` like operation which doesn't need a GPU kernel. Will know more once I can run. The trace is super helpful to have though.

**Comment by parthmannan** [comment]
Looks like it is not Thunder that's causing the slowdown. Removing `thunder.jit` once the above patch is applied results in no noticeable slowdown.
The slowdown comes from the patch which disables the APEX GroupNorm and replaces it with torch GroupNorm. The APEX kernel is far more optimized and given that GroupNorm is a sizeable portion of UNet in StableDiffusion, the slowdown makes sense.
To retain similar speedup in Thunder, we need to either a) match the APEX GroupNorm performance using nvFuser or b) Pattern match `torch.nn.GroupNorm` to APEX GroupNorm.

**Comment by tfogal** [comment]
> Looks like it is not (exactly) Thunder ...

Thanks so much for your analysis, Parth! This is super helpful.

> The slowdown comes from the patch which disables the APEX GroupNorm and replaces it with torch GroupNorm.

Ahh, I imagine this was put in because thunder broke on the GroupNorm operator? @athitten can you confirm?

> or b) Pattern match torch.nn.GroupNorm to APEX GroupNorm. or c) Register APEX GroupNorm so that Thunder can trace through it. (probably easiest)

For this option, is it single-op -> single-op pattern matching? I ask because pattern matching a single operator happens to be a subcase of pattern matching that could be implemented much more easily / could be done quickly with an ad hoc solution.
But I agree (c) is the best route forward, just curious.

**Comment by mruberry** [comment]
triage review —

- @tfogal if groupnorm from apex is the truth, would you file an issue to add apex groupnorm to the apex executor and create a benchmark for groupnorm?
- you can take over this issue to do this

**Comment by parthmannan** [comment]
> For this option, is it single-op -> single-op pattern matching?
Yes this is matching torch.nn.GroupNorm op to APEX GroupNorm function and can be an option for us. Ideally, we do both b) and c) so that we can understand any code using APEX GroupNorm already and also replace torch GroupNorm with APEX GroupNorm automatically.

**Comment by tfogal** [comment]
Thanks, Parth.

Editing title and the original issue comment to reflect triage discussion: we need to implement GroupNorm in the APEX executor.

**Comment by athitten** [comment]
> Ahh, I imagine this was put in because thunder broke on the GroupNorm operator? @athitten can you confirm?

Yes torch.nn.GroupNorm was put in because thunder broke on the GroupNorm operator. 
Using the torch GroupNorm operator in both cases and comparing with thunder.jit and without thunder.jit, I see thunder.jit being slower by 30 ms per iteration. Is this slowdown okay @tfogal @parthmannan ?

**Comment by parthmannan** [comment]
@athitten Can you post the logs where you see 30ms slowdown? Is this for a single iteration?
In my tests on H100, the entire iteration is 50ms and the slowdown is on the order ~1-1.5 ms.

**Comment by athitten** [comment]
Hi @parthmannan here are the logs with thunder.jit and without for FP32: 
[autoencoder_with_thunder_fp32.log](https://github.com/user-attachments/files/15569075/autoencoder_with_thunder_fp32.log)
[autoencoder_wo_thunder_fp32.log](https://github.com/user-attachments/files/15569076/autoencoder_wo_thunder_fp32.log)
I am running on a single A100. The difference is around 20 ms for `FP32`.

However with `BF16`, `thunder.jit` is slower by around 37ms. Here are the logs for `BF16`: 
[autoencoder_with_thunder_jit_bf16.log](https://github.com/user-attachments/files/15569121/autoencoder_with_thunder_jit_bf16.log)
[autoencoder_wo_thunder_jit_bf16.log](https://github.com/user-attachments/files/15569125/autoencoder_wo_thunder_jit_bf16.log)

**Comment by parthmannan** [comment]
@athitten Yea, I think you are right. I was able to test on A6000 and Thunder is slower even without the APEX GroupNorm. I need to dig further for that.
Can you tell me the patch you used to generate the thunder trace? Where in the NeMo code do we insert the Thunder trace calls? That'll be helpful.

**Comment by parthmannan** [comment]
I think I kind of know where the problem is but we need better understanding of NeMo code to solve this. Thunder is running the Encoder in FP32 even when precision is set to BF16. Even without Thunder, the model gets the input in FP32 but it converts it to BF16 before running GEMMs but Thunder GEMMs are running everything in FP32.
Where in NeMo is the precision set and where is AMP being taken care of? Seems like it doesn't kick in when using Thunder.

**Comment by IvanYashchuk** [comment]
> I think I kind of know where the problem is but we need better understanding of NeMo code to solve this. Thunder is running the Encoder in FP32 even when precision is set to BF16. Even without Thunder, the model gets the input in FP32 but it converts it to BF16 before running GEMMs but Thunder GEMMs are running everything in FP32. Where in NeMo is the precision set and where is AMP being taken care of? Seems like it doesn't kick in when using Thunder.

Thunder should recognize PyTorch's autocast context and transform the initial trace accordingly. Here's an example:
```py
In [1]: import thunder

In [2]: import torch

In [3]: a = torch.randn((3, 3), device="cuda")

In [4]: @thunder.jit
   ...: def func(a, b): return a @ b

In [5]: with torch.autocast("cuda", dtype=torch.bfloat16):
   ...:     out = func(a, a)
   ...: 

In [6]: out.dtype
Out[6]: torch.bfloat16

In [7]: thunder.last_traces(func)[-1]
Out[7]: 
# Constructed by Delete Last Used (took 0 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, b):
  # a: "cuda:0 f32[3, 3]"
  # b: "cuda:0 f32[3, 3]"
  [t0, t1] = nvFusion0(a, b)
    # t0 = prims.convert_element_type(a, dtypes.bfloat16)  # t0: "cuda:0 bf16[3, 3]"
    # t1 = prims.convert_element_type(b, dtypes.bfloat16)  # t1: "cuda:0 bf16[3, 3]"
  del a, b
  t2 = torch.matmul(t0, t1)  # t2: "cuda:0 bf16[3, 3]"
    # t2 = ltorch.matmul(t0, t1)  # t2: "cuda:0 bf16[3, 3]"
      # t2 = prims.matmul(t0, t1)  # t2: "cuda:0 bf16[3, 3]"
  del t0, t1
  return t2
```

It's best to print the initial trace, check if this code path is taken https://github.com/Lightning-AI/lightning-thunder/blob/4cc7b64ecbb9b9a28081c09ea00cf61093e57d9b/thunder/__init__.py#L554
and that there are rules for all ops that are expected to be downcasted https://github.com/Lightning-AI/lightning-thunder/blob/4cc7b64ecbb9b9a28081c09ea00cf61093e57d9b/thunder/core/transforms.py#L3685-L3719
